File: src/bindings/c/pressure_resistance_flow.f90
module pressure_resistance_flow_c
implicit none
private

contains



  !
  !###################################################################################
  !
  ! the main growing subroutine. Generates a volume-filling tree into a closed surface.
  subroutine occlusion_list_c(surface_elems_len, surface_elems) bind(C, name="occlusion_list_c")

    !use arrays,only: dp
    !use iso_c_binding, only: c_ptr
    !use utils_c, only: strncpy
    !use other_consts, only: MAX_FILENAME_LEN
    use pressure_resistance_flow,only: occlusion_list
    implicit none

    integer,intent(in) :: surface_elems_len
    integer,intent(in) :: surface_elems(surface_elems_len)

    call occlusion_list(surface_elems)

  end subroutine occlusion_list_c
  !
  !###################################################################################
  !
  ! makes the vessels given rigid.
  subroutine compliance_list_c(surface_elems_len, surface_elems) bind(C, name="compliance_list_c")

    !use arrays,only: dp
    !use iso_c_binding, only: c_ptr
    !use utils_c, only: strncpy
    !use other_consts, only: MAX_FILENAME_LEN
    use pressure_resistance_flow,only: compliance_list
    implicit none

    integer,intent(in) :: surface_elems_len
    integer,intent(in) :: surface_elems(surface_elems_len)

    call compliance_list(surface_elems)

  end subroutine compliance_list_c

!!!###################################################################################

subroutine evaluate_prq_c(mesh_type,mesh_type_len,vessel_type,vessel_type_len,grav_dirn,grav_factor,bc_type,bc_type_len,inlet_bc, &
               outlet_bc,remodeling_grade) bind(C, name="evaluate_prq_c")

  use iso_c_binding, only: c_ptr
  use utils_c, only: strncpy
  use other_consts, only: MAX_STRING_LEN
  use arrays, only: dp
  use pressure_resistance_flow, only: evaluate_prq
  implicit none

  type(c_ptr), value, intent(in) :: mesh_type,bc_type,vessel_type
  integer,intent(in) :: mesh_type_len,bc_type_len, vessel_type_len,grav_dirn
  character(len=MAX_STRING_LEN) :: mesh_type_f,bc_type_f,vessel_type_f
  real(dp),intent(in) :: grav_factor,inlet_bc,outlet_bc,remodeling_grade

  call strncpy(mesh_type_f, mesh_type, mesh_type_len)
  call strncpy(bc_type_f, bc_type, bc_type_len)
  call strncpy(vessel_type_f, vessel_type, vessel_type_len)

  call evaluate_prq(mesh_type_f,vessel_type_f,grav_dirn,grav_factor,bc_type_f,inlet_bc,outlet_bc,remodeling_grade)

end subroutine evaluate_prq_c

!###################################################################################
end module pressure_resistance_flow_c

File: src/bindings/c/pressure_resistance_flow.c
#include "pressure_resistance_flow.h"
#include <string.h>

void compliance_list_c(int *elemlist2_len, int elemlist2[]);
void occlusion_list_c(int *elemlist_len, int elemlist[]);
void set_perfusion_flows_c(int *element_ids_len, int element_ids[], int *flow_values_len, double flow_values[]);
void clear_perfusion_flows_c(void);

void evaluate_prq_c(const char *mesh_type, int *mesh_type_len, const char *vessel_type, int *vessel_type_len,int *grav_dirn, double *grav_factor, const char *bc_type, int *bc_type_len,  double *inlet_bc, double *outlet_bc,  double *remodeling_grade);

void compliance_list(int elemlist2_len, int elemlist2[])
{
  compliance_list_c(&elemlist2_len, elemlist2);
}

void occlusion_list(int elemlist_len, int elemlist[])
{
  occlusion_list_c(&elemlist_len, elemlist);
}

void set_perfusion_flows(int element_ids_len, int element_ids[], int flow_values_len, double flow_values[])
{
  set_perfusion_flows_c(&element_ids_len, element_ids, &flow_values_len, flow_values);
}

void clear_perfusion_flows(void)
{
  clear_perfusion_flows_c();
}

void evaluate_prq(const char *mesh_type, const char *vessel_type,int grav_dirn, double grav_factor, const char *bc_type, double inlet_bc, double outlet_bc, double remodeling_grade)
{
    int mesh_type_len = strlen(mesh_type);
	int bc_type_len = strlen(bc_type);
	int vessel_type_len = strlen(vessel_type);
    evaluate_prq_c(mesh_type, &mesh_type_len,vessel_type, &vessel_type_len, &grav_dirn, &grav_factor, bc_type, &bc_type_len, &inlet_bc, &outlet_bc, &remodeling_grade);
}
